src/main.rs

```diff
@@ -1,126 +1,267 @@
+                // Unsafe because we are guaranteeing that these indexes are not out of bounds
+                let color = unsafe { color.get([x, y]) };
+                let rgba = bgra_u32_to_rgba(*color);
+
+                let pixel = img.get_pixel_mut(
+                    (column * image_width + x) as u32,
+                    (row * image_height + y) as u32,
+                );
+
+                *pixel = vek_rgba_to_image_rgba(rgba);
+            }
+        }
+
+        color.clear(0);
+        depth.clear(1.0);
     }
-}
 
-//TODO: Read the modules chapter and split this file up
-
-/// The ID of a vertex within a specific mesh
-/// Only to be used for the mesh where it came from
-#[derive(Debug, Clone)]
-struct VertexId(usize);
-/// The ID of a edge within a specific mesh
-/// Only to be used for the mesh where it came from
-#[derive(Debug, Clone)]
-struct EdgeId(usize);
-/// The ID of a face within a specific mesh
-/// Only to be used for the mesh where it came from
-#[derive(Debug, Clone)]
-struct FaceId(usize); //TODO: lookup "rust newtype pattern"
-//TODO: Add good docs to everything
-#[derive(Debug, Clone)]
-struct MeshId(usize); //TODO: index into scene mesh array?
-
-#[derive(Debug, Clone)]
-struct Mesh {
-    // Used to associate with other things
-    id: MeshId, //TODO: Figure out the right type for this
-    vertices: Vec<Point2D>,
-    /// List of edge indexes (a, b) == (b, a)
-    edges: Vec<(VertexId, VertexId)>,
-    faces: Vec<Vec<VertexId>>,
-    //TODO: Associate vertices with material IDs
+    img.save("spritesheet.png").expect("unable to write image");
 }
 
-impl Mesh {
-    pub fn vertex(&self, vid: VertexId) -> Point2D {
-        unimplemented!();
-    }
+fn preview_window(
+    image_width: usize,
+    image_height: usize,
+    model: Mat4<f32>,
+    view: Mat4<f32>,
+    projection: Mat4<f32>,
+    frames: &[Vec<Mesh>],
+) {
+    let scale = 16;
+    let background = rgba_to_bgra_u32(Rgba {r: 0.62, g: 0.62, b: 0.62, a: 1.0});
 
-    pub fn edge(&self, eid: EdgeId) -> (Point2D, Point2D) {
-        unimplemented!();
-    }
+    let mut color = Buffer2d::new([image_width, image_height], background);
+    let mut depth = Buffer2d::new([image_width, image_height], 1.0);
 
-    pub fn face(&self, fid: usize) -> Vec<Point2D> {
-        unimplemented!();
-    }
+    // Scaled screen buffer
+    let (screen_width, screen_height) = (image_width * scale, image_height * scale);
+    let mut screen = Buffer2d::new([screen_width, screen_height], 0);
 
-    //TODO: Lookup "impl trait" syntax and learn about iterators
-    pub fn vertices(&self) -> impl Iterator<Item=(VertexId, Point2D)> {
-        unimplemented!();
-    }
+    let (axis_width, axis_height) = (128, 128);
+    let axis = if screen_width > axis_width && screen_height > axis_height {
+        let mut axis_color = Buffer2d::new([axis_width, axis_height], background);
+        render_axis(&mut axis_color, view);
+        Some(axis_color)
+    } else { None };
```

Comment:
What was the reasoning behind setting the axis dimensions to this value 😮  What if we assign the axis width and height based on a percentage of the screen dimensions?
I think the reason why I couldn't see axis rendered on my screen on my windows laptop was because of this if statement here.

Reply:
Yup that can work too. Frankly, we should just use the minifb API better. They have a [Scale enum](https://docs.rs/minifb/0.11.2/minifb/enum.Scale.html) that we could use. My screen is high DPI and yours might be even higher :open_mouth: so we can set the scale factor accordingly.

That can be future work.